### PR TITLE
Update PolyPartition / Triangulator library

### PR DESCRIPTION
**Pull request by aaronfranke**
This PR updates the [PolyPartition](https://github.com/ivanfratric/polypartition) library (previously the files were called "triangulator") to commit [7bdffb4](https://github.com/ivanfratric/polypartition/commit/7bdffb428b2b19ad1c43aa44c714dcc104177e84).

The only difference between the version in the repo and the version in this PR is that this PR uses Godot's types (`List`, `Set`, `Vector`, `Vector2`) instead of the standard library (`std::list`, `std::set`, `std::vector`) and PolyPartition's `TPPLPoint`.

### CODE REVIEW

**Comment by akien-mga** [comment]
> The only difference between the version in the repo and the version in this PR is that this PR uses Godot's types (`List`, `Set`, `Vector`, `Vector2`) instead of the standard library (`std::list`, `std::set`, `std::vector`) and PolyPartition's `TPPLPoint`.

Why? We shouldn't modify thirdparty code if we don't have to, and that's a very intrusive change to thirdparty code, making it difficult to do any future update.

**Comment by aaronfranke** [comment]
@akien-mga Well, because the existing version in Godot was using Godot's types, so I kept that part the same.

It should be easy to change it back to using the standard library by just copying the upstream version again, though in the case of `TPPLPoint` I think it makes sense to keep using Godot's `Vector2` since it's a drop-in replacement with one typedef.

**Comment by akien-mga** [comment]
I see. Then I guess it's fine for this PR, but yeah I think a follow-up could try to undo all these Godot-specific changes and use the unmodified thirdparty code. If we need to write a `Vector2` wrapper around `TPPLPoint` that can be done too.

Could you update `thirdparty/README.md` with the rename and the actual version documentation, now that it's known?

**Comment by akien-mga** [comment]
Could you also add a `.patch` file that documents the changes made to change from STL to Godot types? You can do it by backing up your modified files, adding the unmodified upstream code on top, `git add` it, then re-copy your modified files on top and `git diff > thirdparty/misc/patches/polypartition-godot-types.patch`.

**Comment by aaronfranke** [comment]
I will do that, keeping a `.patch` file is a good idea.

I looked into what it would take to keep `std::`. We would have to update the Godot code that depends on this to use the standard library, and add conversions to and from Godot types for the input and output of methods such as `Geometry2D::decompose_polygon_in_convex`, so I think that's not a good idea.

**Comment by Xrayez** [comment]
In Goost, I've exposed quite a bunch of hidden PolyPartition functionality, and discovered that there's actually a hidden bug in `Triangulate_MONO` (with Godot types): https://github.com/goostengine/goost/blob/c5ef6734b87057f0d73ca0fcbd6c3a50c862480c/core/math/2d/geometry/poly/decomp/polypartition/poly_decomp_polypartition.cpp#L94-L98

This is also something which #16423 proposed to enable over ear-clipping which is currently the default (would mostly avoid "Bad Polygon!" errors while drawing polygons).

So, I'm just saying that I'd be careful modifying thirdparty code, and yeah I'd certainly prefer to undo those Godot-specific changes because there's a high chance of breaking things that way, especially when the library is updated.

> add conversions to and from Godot types for the input and output of methods such as `Geometry2D::decompose_polygon_in_convex`, so I think that's not a good idea.

That's how Clipper library is currently used in Godot for polygon boolean methods (mainly due to the way how points needs to be scaled up and down to achieve robust computations), and it's quite fast even with conversion going on.

This is something which can be done via dedicated PR though.

**Comment by aaronfranke** [comment]
@Xrayez If you would like to add onto this PR, that would be nice too. It would result in fewer diffs than if this PR was accepted as it is now and then these files were changed and the patch deleted.

**Comment by Xrayez** [comment]
I'd certainly work on this but unfortunately I'm not using Godot 4.0 currently so it seems like such a low-priority task to be honest... I think it's fine to have those Godot types for now. Perhaps this update may fix the crash with `Triangulate_MONO` and this wouldn't be critical any longer. It's just something to be ever of.

File: thirdparty/README.md [hunk 1]
```diff
@@ -441,9 +441,9 @@ Collection of single-file libraries used in Godot components.
   * Upstream: https://github.com/nothings/stb
   * Version: 1.20 (314d0a6f9af5af27e585336eecea333e95c5a2d8, 2020)
   * License: Public Domain or Unlicense or MIT
-- `triangulator.{cpp,h}`
-  * Upstream: https://github.com/ivanfratric/polypartition (`src/polypartition.cpp`)
-  * Version: TBD, class was renamed
+- `polypartition.{cpp,h}`
+  * Upstream: https://github.com/ivanfratric/polypartition (`src/polypartition.{cpp,h}`)
+  * Version: 7bdffb428b2b19ad1c43aa44c714dcc104177e84 (2021, with patch for Godot types)
```
Comment (akien-mga) [inline thirdparty/README.md:446]: Should be:
```suggestion
  * Version: git (7bdffb428b2b19ad1c43aa44c714dcc104177e84, 2021)
  * Modifications: Change from STL to Godot types (see provided patch).
```

The entry should also be moved further up, they're ordered alphabetically.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**